### PR TITLE
Add `freeable`, to identify packages that may be freed without downgrading

### PR DESCRIPTION
**Pull request by timholy**
Moved from https://github.com/JuliaLang/julia/pull/16312.

I also took the liberty to fix `String` deprecations, hope that's OK.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,3 +30,15 @@ Keyword parameters:
 
 ### license([lic])
 List all bundled licenses. If a license label specified as a parameter then a full text of the license will be printed.
+
+### freeable([io])
+Returns a list of packages which are good candidates for
+`Pkg.free`. These are packages for which you are not tracking the
+tagged release, but for which a tagged release is equivalent to the
+current version. You can use `Pkg.free(Pkg.freeable())` to
```
Comment (tkelman) [inline README.md]: `Pkg.free(PkgDev.freeable())`

**Comment by timholy** [comment]
OK to merge, @wildart?

File: src/entry.jl [hunk 2]
```diff
@@ -286,4 +287,44 @@ function check_metadata(pkgs::Set{ByteString} = Set{ByteString}())
     return
 end
 
+function freeable(io::IO = STDOUT)
+    function latest_sha(pkgname)
+        avail = Read.available(pkgname)
+        k = sort(collect(keys(avail)))
+        isempty(k) ? nothing : avail[k[end]]
```
Comment (tkelman) [inline src/entry.jl]: I'd say use a Nullable if performance was an issue

File: src/entry.jl [hunk 3]
```diff
@@ -8,6 +8,7 @@ using Base.Pkg.Types
 import ..PkgDev
 import ..PkgDev.GitHub
 using ..PkgDev: getrepohttpurl
+using Compat
```
Comment (yuyichao) [inline src/entry.jl]: This is not actually used?

File: src/entry.jl [hunk 4]
```diff
@@ -286,4 +287,44 @@ function check_metadata(pkgs::Set{ByteString} = Set{ByteString}())
...
+    end
+    freelist = Any[]
+    cd(Pkg.dir()) do
+        instd = Read.installed()
+        first = true
+        k, v = collect(keys(instd)), collect(values(instd))
+        p = sortperm(k)
+        k = k[p]
+        v = v[p]
+        for (pkg,status) in zip(k,v)
```
Comment (tkelman) [inline src/entry.jl]: how about just `for (pkg, status) in sort!(collect(Read.installed()), by=first)` ?

File: src/entry.jl [hunk 5]
```diff
@@ -286,4 +287,44 @@ function check_metadata(pkgs::Set{ByteString} = Set{ByteString}())
     return
 end
 
+function freeable(io::IO = STDOUT)
+    function latest_sha(pkgname)
```
Comment (tkelman) [inline src/entry.jl]: this is "latest tag" right?

File: src/entry.jl (hunk 3 above)
Comment (wildart) [inline src/entry.jl]: not really

Comment (tkelman) [inline src/entry.jl]: may be useful later on in case PkgDev is needed to bisect on issues in Base

Comment (yuyichao) [inline src/entry.jl]: Assuming this is for better bisecting of PkgDev without having String depwarns, this should be `import Compat.String` instead.

Comment (tkelman) [inline src/entry.jl]: More for the bisecting of Base without changing PkgDev - if you don't use Compat that could break entirely.

Comment (yuyichao) [inline src/entry.jl]: > More for the bisecting of Base without changing PkgDev

Yeah, that's what I mean too. The point is though, this `using Compat` is **not** doing that, unless this is not for the String change.

**Comment by timholy** [comment]
Thanks for the review comments, everyone. Much nicer now. I also added precompilation (saves 2-3x, depending on whether Comapt is already loaded).

**Comment by yuyichao** [comment]
Can you add the missing `import Compat.String`?

**Comment by timholy** [comment]
Already there (maybe hit refresh?)

**Comment by yuyichao** [comment]
The one in `src/entry.jl` where Tony and I were commenting on. Not the one in `src/PkgDev.jl`

**Comment by timholy** [comment]
We are currently experience technical difficulties. Please be patient while we wait for our brain to come back on line.

**Comment by yuyichao** [comment]
LGTM (at least the string part =) )

**Pull request merged**